New Module Submision DellEMC PowerMax Modules

rawstorage opened the pull request:
SUMMARY
Added new storage module for PowerMAX, VMAX All Flash and VMAX3 arrays.

ISSUE TYPE
New Module Pull Request
COMPONENT NAME
Storage Modules for PowerMAX, VMAX3 and VMAX all Flash

Storage Provisioning Modules Created and tested
dellpmax_createsg.py

rawstorage committed:
Adding DellEMC Modules for VMAX and PowerMax Storage Operations

rawstorage committed:
Adding DellEMC Modules for VMAX and PowerMax Storage Operations
Basic Provisioning, lun expansion, remote replication control and
snapshot creation and manipulation modules tested and submitted.

rawstorage committed:
Adding DellEMC Modules for VMAX and PowerMax Storage Operations
Basic Provisioning, lun expansion, remote replication control and
snapshot creation and manipulation modules tested and submitted.

added __init__.py for compliance with module submission rules

rawstorage committed:
Adding DellEMC Modules for VMAX and PowerMax Storage Operations
Basic Provisioning, lun expansion, remote replication control and
snapshot creation and manipulation modules tested and submitted.

added __init__.py for compliance with module submission rules

rawstorage committed:
Adding DellEMC Modules for VMAX and PowerMax Storage Operations
Basic Provisioning, lun expansion, remote replication control and
snapshot creation and manipulation modules tested and submitted.

added __init__.py for compliance with module submission rules

rawstorage committed:
removing additional files, will be readded after initial push

rawstorage committed:
PEP8 formatting and moved imports to follow examples

rawstorage committed:
corrected documentation comments and example to be valid YML

rawstorage committed:
removed trailing whitespace and updated GPL license

rawstorage committed:
Updated Documentation YML format removing linebreaks and spaces.

rawstorage committed:
Updated Documentation YML format removing linebreaks and spaces. Corrected indentation on documentation to remove duplicate key references

rawstorage committed:
verified documentation is now correct with ansible-doc -t module dellpmax_createsg
module should now pass.

rawstorage committed:
import statment moved to try/except, cleaned up whitespace and other
pep8 changes

rawstorage committed:
corrected documentation and choices

rawstorage committed:
Fixed Trailing whitespaces, and documentation type

rawstorage committed:
Fixed Trailing whitespaces, and documentation type

rawstorage committed:
corrected _ and removed whitepace in examples

rawstorage committed:
added dellpmax_addvolume.py

rawstorage committed:
added dellpmax_addvolume.py

rawstorage committed:
added dellpmax_addvolume.py

rawstorage committed:
removed reserved words and trailing whitespace.

rawstorage committed:
added createhost, createhostgroup, create portgroup, createmaskingview

rawstorage committed:
added createhost, createhostgroup, create portgroup, createmaskingview

rawstorage committed:
corrected pep8 issues

rawstorage committed:
corrected pep8 issues and arguments on documentation

rawstorage committed:
added dellpmax.py to module_utils

rawstorage committed:
Removed bland line for pep8

rawstorage committed:
renamed dellpmax.py in moudle_utils to dellemc.py

rawstorage committed:
renamed file to dellemc_pmax_createsg.py for submission to ansible.

rawstorage committed:
removed other modules for single module submission

rawstorage commented:
@Akasurde hopefully this is good to go now, let me know if there is any issue

Akasurde commented on lib/ansible/module_utils/dellemc.py:
```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
+
```
Replace line 2 to 29 with - 

```
# Copyright: (c) 2018, Paul Martin <paule.marting@dell.com>
# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
```

rawstorage replied:
Done

Akasurde commented on lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py:
```diff
@@ -0,0 +1,200 @@
+#!/usr/bin/python
+# Copyright (C) 2018 DellEMC
```
```suggestion
# Copyright: (C) 2018, DellEMC
```

rawstorage replied:
Accepted change

Akasurde commented on lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py:
```diff
@@ -0,0 +1,200 @@
+version_added: "2.8"
+description:
+  - "This module has been tested against UNI 9.0. Every effort has been made
+  to verify the scripts run with valid input. These modules are a tech preview"
```
```suggestion
  to verify the scripts run with valid input. These modules are a tech preview."
```

Akasurde commented on lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py:
```diff
@@ -0,0 +1,200 @@
+      - CYL
+    description:
+      - "String value, default is set to GB"
+    required: false
```
```suggestion
    required: false
    default: GB
```

Akasurde commented on lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py:
```diff
@@ -0,0 +1,200 @@
+      - "integer value for the number of volumes. Minimum is 1, module will
+      fail if less than one volume is specified or value is 0. If volumes are
+      required of different sizes, additional tasks should be added to
+      playbooks to use dellemc_pmax_addvolume module"
```
```suggestion
      playbooks to use M(dellemc_pmax_addvolume) module."
```

Akasurde commented on lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py:
```diff
@@ -0,0 +1,200 @@
+  sgname:
+    description:
+      - "Storage Group name 32 Characters no special characters other than
+      underscore"
```
```suggestion
      underscore."
```

Akasurde commented on lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py:
```diff
@@ -0,0 +1,200 @@
+    description:
+      - "Storage Group name 32 Characters no special characters other than
+      underscore"
+    required: true
```
```suggestion
    required: true
    type: str
```

rawstorage replied:
committed change and reviewing to make similar for other arguments.

Akasurde commented on lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py:
```diff
@@ -0,0 +1,200 @@
+    description:
+      - "Storage Resource Pool Name, Default is set to SRP_1, if your system
+      has mainframe or multiple pools you can set this to a different value to
+      match your environment"
```
```suggestion
      match your environment."
```

Akasurde commented on lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py:
```diff
@@ -0,0 +1,200 @@
+      - "Storage Resource Pool Name, Default is set to SRP_1, if your system
+      has mainframe or multiple pools you can set this to a different value to
+      match your environment"
+    required: false
```
Specify type for all parameters like 

```suggestion
    required: false
    type: str
```

Akasurde commented on lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py:
```diff
@@ -0,0 +1,200 @@
+    description:
+      - "Integer, version of unipshere software  e.g. 90"
+    required: true
+  verifycert:
```
There is a parameter called `validate_certs` which used other modules. It would be great if you could that here so it will match every other module. [Example](https://github.com/ansible/ansible/blob/013c42b14fa06ff6355d8a9d52430f4cb1262047/lib/ansible/utils/module_docs_fragments/vmware.py#L30)

rawstorage replied:
Akasurde, 
the verify_cert parameter here is specific to the PyU4V package I am importing,.  I can't change this one unfortunately.

Akasurde commented on lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py:
```diff
@@ -0,0 +1,200 @@
+        workload: None
+        volumeIdentifier: 'REDO'
+'''
+RETURN = '''
```
RETURN is optional, but useful end user to understand what module returns.

rawstorage replied:
I will change module to return a dictionary describing the storage group that has been created.

rawstorage replied:
added sample return for successful run
dellemc_pmax_createsg:
    description: Information about storage group created
    returned: success 
    type: dict
    sample: '{
        "storagegroup_detail": {
            "VPSaved": "100.0%",
            "base_slo_name": "Diamond",
            "cap_gb": 1.0,
            "compression": true,
            "device_emulation": "FBA",
            "num_of_child_sgs": 0,
            "num_of_masking_views": 0,
            "num_of_parent_sgs": 0,
            "num_of_snapshots": 0,
            "num_of_vols": 1,
            "service_level": "Diamond",
            "slo": "Diamond",
            "slo_compliance": "STABLE",
            "srp": "SRP_1",
            "storageGroupId": "Ansible_SG",
            "type": "Standalone",
            "unprotected": true,
            "vp_saved_percent": 100.0
            }
        }'
'''

Akasurde commented on lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py:
```diff
@@ -0,0 +1,200 @@
+                'via PIP')
+        module.exit_json(changed=changed)
+
+    conn = PyU4V.U4VConn(server_ip=module.params['unispherehost'], port=8443,
```
If this is going to be common to other DellEMC modules then you can move this method to `module_utils/dellemc.py`.

rawstorage replied:
Implemented reccomendation

Akasurde commented on lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py:
```diff
@@ -0,0 +1,200 @@
+            unispherehost=dict(required=True),
+            universion=dict(type='int', required=False),
+            verifycert=dict(type='bool', required=True),
+            user=dict(type='str', required=True),
```
If user, password, parameters are common to all other modules then it is better to create common Argument spec like [this](https://github.com/ansible/ansible/blob/013c42b14fa06ff6355d8a9d52430f4cb1262047/lib/ansible/module_utils/vmware.py#L448) one

rawstorage replied:
I'll rework these changes and push once retested.

Akasurde reviewed (changes requested):
Are you planning to add test case for this change ?

AnilDegwekar commented on lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py:
```diff
@@ -0,0 +1,200 @@
+                                     )
+        changed = True
+    else:
+        module.fail_json(msg='Storage Group Already Exists')
```
As per the 'idempotency' principle of Ansible, here we need to check if the existing Storage Group matches the one that is requested. If so, should return success rather than a failure.

rawstorage replied:
@AnilDegwekar so if I change the else statement here to exit the module with changed=false and still success this should be ok?

rawstorage replied:
@AnilDegwekar looking at this again, my intent here is to fail if the storage group exists.  I really dont' want to be able to continue if the storage group name already exists as this could lead to a masking view being created later presenting storage volumes to multiple hosts causing data loss.

AnilDegwekar replied:
From Ansible documentation:

Modules should be idempotent, that is, running a module multiple times in a sequence should have the same effect as running it just once. One way to achieve idempotency is to have a module check whether its desired final state has already been achieved, and if that state has been achieved, to exit without performing any actions. If all the modules a playbook uses are idempotent, then the playbook itself is likely to be idempotent, so re-running the playbook should be safe.

AnilDegwekar replied:
https://docs.ansible.com/ansible/latest/user_guide/playbooks_intro.html

Akasurde committed:
Update lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py

Co-Authored-By: rawstorage <paule.martin@dell.com>

Akasurde committed:
Update lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py

Co-Authored-By: rawstorage <paule.martin@dell.com>

Akasurde committed:
Update lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py

Co-Authored-By: rawstorage <paule.martin@dell.com>

Akasurde committed:
Update lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py

Co-Authored-By: rawstorage <paule.martin@dell.com>

Akasurde committed:
Update lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py

Co-Authored-By: rawstorage <paule.martin@dell.com>

Akasurde committed:
Update lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py

Co-Authored-By: rawstorage <paule.martin@dell.com>

Akasurde committed:
Update lib/ansible/modules/storage/dellemc/dellemc_pmax_createsg.py

Co-Authored-By: rawstorage <paule.martin@dell.com>

rawstorage committed:
added global argument spec and pmaxapi module to be reused across
modules

rawstorage committed:
Merge remote-tracking branch 'origin/devel' into devel

rawstorage committed:
merged changes from Anisble team and made suggested changes

rawstorage committed:
Pep8 issues on commit, added spaces on variable assignment and removed
trailing lines in dellemc.py

Set Default value for GB on

rawstorage commented:
@Akasurde All issues should be resolved except for a return value..  While I agree a return is preferable in a lot of instances in this case it's not really required, the success status will ultimately drive on the next task which would be to present the storage group to a host.  Hoping this is enough for successful submission.  I've checked other storage vendors equivalent modules and there is precedence for not having a return, hopefully mine can be accepted on same basis.  Thanks again for the patience in reviewing this module. It will make the submission of subsequent modules much simpler. I'

rawstorage committed:
added sample return,

rawstorage committed:
removed trailing whitespace line 144

rawstorage committed:
removed trailing whitespace line 144

Akasurde commented:
Closing and re-opening for CI trigger.

rawstorage committed:
added full stop to comment on line 194 to trigger CI

rawstorage commented:
@Akasurde can you recheck this, I've addressed all the comments.  I don't plan on doing unit tests for these modules unless expressly required.  We have Unit Tests for all modules in  PyU4V.  I have however retested all changes prior to submission to go though fail and pass scenarios.  I have more modules ready to submit in the new format using the common functions in dellemc.py in module_utils.  Thanks again for all the pointers on this.  I'm getting much better at spotting my own pep8 errors and have my whitespace highlighted in my IDE making it easier.

rawstorage commented:
Closing Pull Request.. Just found out there is joint effort in progress that will push a bunch of DellEMC Anisble modules for Mulitple DELL EMC arrays this project would clash.
